File: easy/ts/383.ts
// https://leetcode.com/problems/ransom-note/?envType=study-plan-v2&envId=top-interview-150
// Time: O(n)
// Space: O(n)
function canConstruct(ransomNote: string, magazine: string): boolean {
  let magCharMap = {};

  for (let i = 0; i < magazine.length; i++) {
    if (magCharMap[magazine[i]]) {
      magCharMap[magazine[i]]++;
    } else {
      magCharMap[magazine[i]] = 1;
    }
  }

  for (let i = 0; i < ransomNote.length; i++) {
    if (magCharMap[ransomNote[i]]) {
      magCharMap[ransomNote[i]]--;
    } else {
      return false;
    }
  }

  return true;
}

File: easy/python/383.py
# https://leetcode.com/problems/ransom-note/?envType=study-plan-v2&envId=top-interview-150
# Time and Space: O(n)
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_map = {}

        for ele in magazine:
            if ele in magazine_map:
                magazine_map[ele] += 1
            else:
                magazine_map[ele] = 1

        for ele in ransomNote:
            if ele in magazine_map and magazine_map[ele] > 0:
                magazine_map[ele] -= 1
            else:
                return False
        return True
